src/test/suite/mdbExtensionController.test.ts

```diff
@@ -1217,38 +1225,317 @@ suite('MDBExtensionController Test Suite', function () {
     );
+
+    await vscode.commands.executeCommand('mdb.saveDocumentToMongoDB');
+
+    assert(mockDocument.name === 'something sweet');
+
+    const expectedMessage =
+      "The document was saved successfully to 'waffle.house'";
+
+    assert(
+      fakeShowInformationMessage.firstArg === expectedMessage,
+      `Expected an error message "${expectedMessage}" to be shown when attempting to add a database to a not connected connection found "${fakeShowInformationMessage.firstArg}"`
+    );
+
+    await fse.remove(mockOpenTextDocument.firstArg.path);
+  });
+
+  test('if the active editor is missing, the save function does not call findOneAndReplace', async () => {
+    const mockDocument = {
+      _id: 'pancakes',
+      name: ''
+    };
+
+    sandbox.replaceGetter(vscode.window, 'activeTextEditor', () => null);
+
+    const mockGetActiveDataService = sinon.fake.returns({
+      findOneAndReplace: (
+        namespace: string,
+        filter: object,
+        replacement: object,
+        options: object,
+        callback: (error: Error | undefined, result: object) => void
+      ) => {
+        mockDocument.name = 'something sweet';
+        callback(undefined, mockDocument);
+      }
+    });
+    sinon.replace(
+      mdbTestExtension.testExtensionController._connectionController,
+      'getActiveDataService',
+      mockGetActiveDataService
+    );
+
+    await vscode.commands.executeCommand('mdb.saveDocumentToMongoDB');
+
+    assert(mockDocument.name === '');
+  });
+
+  test("json files that are not MongoDB documents won't be saved to database", async () => {
+    const mockDocument = {
+      _id: 'pancakes',
+      name: ''
+    };
+
+    sandbox.replaceGetter(vscode.window, 'activeTextEditor', () => ({
+      document: {
+        uri: {
+          query: ''
+        }
+      }
+    }));
+
+    const mockGetActiveDataService = sinon.fake.returns({
+      findOneAndReplace: (
+        namespace: string,
+        filter: object,
+        replacement: object,
+        options: object,
+        callback: (error: Error | undefined, result: object) => void
+      ) => {
+        mockDocument.name = 'something sweet';
+        callback(undefined, mockDocument);
+      }
+    });
+    sinon.replace(
+      mdbTestExtension.testExtensionController._connectionController,
+      'getActiveDataService',
+      mockGetActiveDataService
+    );
+
+    await vscode.commands.executeCommand('mdb.saveDocumentToMongoDB');
+
+    assert(mockDocument.name === '');
+  });
+
+  test("if a user is not connected, documents won't be saved to MongoDB", async () => {
+    const fakeVscodeErrorMessage = sinon.fake();
+    sinon.replace(vscode.window, 'showErrorMessage', fakeVscodeErrorMessage);
+
+    sandbox.replaceGetter(vscode.window, 'activeTextEditor', () => ({
+      document: {
+        uri: {
+          query: [
+            '?documentLocation=mongodb',
+            'namespace=waffle.house',
```

Comment:
I’m not sure if we have to care about this, but some characters like `#` would break URLs but are allowed in collection names … is that an issue here? :smile:

Reply:
Vscode cares about it for us :) we shouldn't do here anything. but what I just noticed is that dataservice fails to save ejson. eg:

```
{
  "_id": {
    "$oid": "5ea8745ee4811fafe8b65ecb"
  },
  "koko": "nothing"
}
```

and throws the `Unable to save document: The dollar ($) prefixed field '$oid' in '_id.$oid' is not valid for storage.` error. Have you ever seen this before or tried to save ejson?

Reply:
I have not seen that error, no – that’s interesting … it sounds like it’s trying to insert the document as-is instead of first deserializing it as EJSON and then inserting it?

Reply:
Found the problem! the one I actually caused 😀 in the try-catch statement above I used JSON.parse() that broke the document format. Replaced this part with EJSON.parse() and now it saves correctly.